src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/NegotiationProtocol.cs

```diff
@@ -31,7 +32,70 @@ public static void WriteMessage(NegotiationMessage negotiationMessage, Stream ou
             TextMessageFormatter.WriteRecordSeparator(output);
         }
 
-        public static bool TryParseMessage(ReadOnlySpan<byte> input, out NegotiationMessage negotiationMessage)
+        public static void WriteResponseMessage(NegotiationResponseMessage negotiationResponseMessage, Stream output)
+        {
+            using (var writer = new JsonTextWriter(new StreamWriter(output, _utf8NoBom, 1024, leaveOpen: true)))
+            {
+                writer.WriteStartObject();
+                if (!string.IsNullOrEmpty(negotiationResponseMessage.Error))
+                {
+                    writer.WritePropertyName(ErrorPropertyName);
+                    writer.WriteValue(negotiationResponseMessage.Error);
+                }
+                writer.WriteEndObject();
+            }
+
+            TextMessageFormatter.WriteRecordSeparator(output);
+        }
+
+        public static bool TryParseResponseMessage(ReadOnlySpan<byte> input, out NegotiationResponseMessage negotiationMessage)
+        {
+            if (!TextMessageParser.TryParseMessage(ref input, out var payload))
+            {
+                throw new InvalidDataException("Unable to parse payload as a negotiation message.");
+            }
+
+            using (var memoryStream = new MemoryStream(payload.ToArray()))
+            {
+                using (var reader = new JsonTextReader(new StreamReader(memoryStream)))
+                {
+                    var token = JToken.ReadFrom(reader);
+                    if (token == null || token.Type != JTokenType.Object)
+                    {
+                        throw new InvalidDataException($"Unexpected JSON Token Type '{token?.Type}'. Expected a JSON Object.");
+                    }
+
+                    var negotiationJObject = (JObject)token;
+                    var error = JsonUtils.GetOptionalProperty<string>(negotiationJObject, ErrorPropertyName);
+                    negotiationMessage = new NegotiationResponseMessage(error);
+                }
+            }
+            return true;
+        }
+
+        public static bool TryParseResponseMessage(ReadOnlySequence<byte> buffer, out NegotiationResponseMessage negotiationMessage, out SequencePosition consumed, out SequencePosition examined)
+        {
+            var separator = buffer.PositionOf(TextMessageFormatter.RecordSeparator);
```

Comment:
This code being identical means we can probably make this a helper. `TryReadTextMessageIntoSingleSpan` or something.